# Web Mining and Applied NLP (44-620)

## Requests, JSON, and NLP

### Student Name: Olakunle Alabi   https://github.com/Tezzyray/json-sentiment

Perform the tasks described in the Markdown cells below.  When you have completed the assignment make sure your code cells have all been run (and have output beneath them) and ensure you have committed and pushed ALL of your changes to your assignment repository.

Make sure you have [installed spaCy and its pipeline](https://spacy.io/usage#quickstart) and [spaCyTextBlob](https://spacy.io/universe/project/spacy-textblob)

Every question that requires you to write code will have a code cell underneath it; you may either write your entire solution in that cell or write it in a python file (`.py`), then import and run the appropriate code to answer the question.

This assignment requires that you write additional files (either JSON or pickle files); make sure to submit those files in your repository as well.

In [40]:
# Create and activate a Python virtual environment. 
# Before starting the project, try all these imports FIRST
# Address any errors you get running this code cell 
# by installing the necessary packages into your active Python environment.
# Try to resolve issues using your materials and the web.
# If that doesn't work, ask for help in the discussion forums.
# You can't complete the exercises until you import these - start early! 
# We also import json and pickle (included in the Python Standard Library).

import json
import pickle
import config
import lyricsgenius
import requests
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

print('All prereqs installed.')
!pip list

All prereqs installed.
Package                Version
---------------------- -----------
anyio                  3.5.0
appdirs                1.4.4
argon2-cffi            21.3.0
argon2-cffi-bindings   21.2.0
asttokens              2.0.5
attrs                  22.1.0
Babel                  2.11.0
backcall               0.2.0
beautifulsoup4         4.11.1
bleach                 4.1.0
blis                   0.7.9
Bottleneck             1.3.5
brotlipy               0.7.0
catalogue              2.0.8
certifi                2022.12.7
cffi                   1.15.1
charset-normalizer     2.0.4
click                  8.1.5
colorama               0.4.6
comm                   0.1.2
conda                  23.1.0
conda-content-trust    0.1.3
conda-package-handling 1.9.0
confection             0.1.0
config                 0.5.1
contourpy              1.0.5
cryptography           38.0.1
cycler                 0.11.0
cymem                  2.0.7
debugpy                1.5.1
decorator              5.1.1

# Question 1

1. The following code accesses the [lyrics.ovh](https://lyricsovh.docs.apiary.io/#reference/0/lyrics-of-a-song/search) public api, searches for the lyrics of a song, and stores it in a dictionary object.  Write the resulting json to a file (either a JSON file or a pickle file; you choose). You will read in the contents of this file for future questions so we do not need to frequently access the API.

In [45]:
import requests
import json

#result = json.loads(requests.get('https://api.lyrics.ovh/v1/They Might Be Giants/Birdhouse in your soul').text)



In [51]:
import lyricsgenius
genius = lyricsgenius.Genius('JtUijw8YGNAomnidRH7gq11zP74lZjoR3lt6LRH2D_8xhk-qcT-SYbsPa8et4zpd')

artist = genius.search_artist("They Might Be Giants", max_songs=5, sort="title")
song = artist.song("Birdhouse in your soul")
artist.save_lyrics("Birdhouse in your soul")

if song is not None:
    lyrics = song.lyrics
    print(f"Lyrics for '{song.title}' by {song.artist}:\n{lyrics}")

Searching for songs by They Might Be Giants...

Song 1: "200 Sbemails (for Homestar Runner)"
Song 2: "2082"
Song 3: "25 O’Clock"
Song 4: "32 Footsteps"
Song 5: "3rd Amendment"

Reached user-specified song limit (5).
Done. Found 5 songs.
Searching for "Birdhouse in your soul" by They Might Be Giants...
Done.
Birdhouse in your soul.json already exists. Overwrite?
(y/n): y
Wrote Birdhouse in your soul.json.
Lyrics for 'Birdhouse in Your Soul' by They Might Be Giants:
52 ContributorsBirdhouse in Your Soul Lyrics[Bridge]
I'm your only friend
I'm not your only friend
But I'm a little glowing friend
But really I'm not actually your friend
But I am

[Chorus]
Blue canary in the outlet by the light switch
Who watches over you
Make a little birdhouse in your soul
Not to put too fine a point on it
Say I'm the only bee in your bonnet
Make a little birdhouse in your soul

[Verse 1]
I have a secret to tell
From my electrical well
It's a simple message and I'm
Leaving out the whistles and bells
So the

# Question 2

2. Read in the contents of your file.  Print the lyrics of the song (not the entire dictionary!) and use spaCyTextBlob to perform sentiment analysis on the lyrics.  Print the polarity score of the sentiment analysis.  Given that the range of the polarity score is `[-1.0,1.0]` which corresponds to how positive or negative the text in question is, do you think the lyrics have a more positive or negative connotaion?  Answer this question in a comment in your code cell.

In [53]:
from textblob import TextBlob

nlp = spacy.load('en_core_web_sm')
nlp.add_pipe("spacytextblob")

def get_sentiment_score(text):
    doc = nlp(text)
    return doc._.polarity

genius = lyricsgenius.Genius('JtUijw8YGNAomnidRH7gq11zP74lZjoR3lt6LRH2D_8xhk-qcT-SYbsPa8et4zpd')


artist = genius.search_artist("They Might Be Giants", max_songs=5, sort="title")


song = artist.song("Birdhouse in your soul")

#Performing sentiment analysis
if song is not None:
    lyrics = song.lyrics
    print(f"Lyrics for '{song.title}' by {song.artist}:\n{lyrics}")

    sentiment_score = get_sentiment_score(lyrics)
    print(f"Sentiment score: {sentiment_score}")
else:
    print("Song not found.")
    
#The sentiment score shows that the lyrics of the songs has negative sentiment score .

Searching for songs by They Might Be Giants...

Song 1: "200 Sbemails (for Homestar Runner)"
Song 2: "2082"
Song 3: "25 O’Clock"
Song 4: "32 Footsteps"
Song 5: "3rd Amendment"

Reached user-specified song limit (5).
Done. Found 5 songs.
Searching for "Birdhouse in your soul" by They Might Be Giants...
Done.
Lyrics for 'Birdhouse in Your Soul' by They Might Be Giants:
52 ContributorsBirdhouse in Your Soul Lyrics[Bridge]
I'm your only friend
I'm not your only friend
But I'm a little glowing friend
But really I'm not actually your friend
But I am

[Chorus]
Blue canary in the outlet by the light switch
Who watches over you
Make a little birdhouse in your soul
Not to put too fine a point on it
Say I'm the only bee in your bonnet
Make a little birdhouse in your soul

[Verse 1]
I have a secret to tell
From my electrical well
It's a simple message and I'm
Leaving out the whistles and bells
So the room must listen to me
Filibuster vigilantly
My name is blue canary
One note spelled l-i-t-e
My st

# Question 3

3. Write a function that takes an artist, song, and filename, accesses the lyrics.ovh api to get the song lyrics, and writes the results to the specified filename.  Test this function by getting the lyrics to any four songs of your choice and storing them in different files.

In [71]:

import lyricsgenius
import json
import os

def get_song_lyrics(artist, song, filename):
    access_token = '6wkG206P0FgF5A4OcBd5LbeG5hxNCQbkGLhKuEB_7Y2iV1jTZT1PckgANAEfPdXk'
    genius = lyricsgenius.Genius(access_token)
    artist = genius.search_artist(artist, max_songs=1, get_full_info=False)
    song = artist.song(song)
    with open(filename+"_lyrics.json", 'w') as filename:
        json.dump(song.lyrics, filename)

get_song_lyrics('The Beatles', 'Hey Jude', 'hey_jude.txt')
get_song_lyrics('Led Zeppelin', 'Stairway to Heaven', 'stairway_to_heaven.txt')
get_song_lyrics('Pink Floyd', 'Another Brick in the Wall', 'another_brick_in_the_wall.txt')
get_song_lyrics('The Rolling Stones', '(I Cant Get No) Satisfaction', '(i_cant_get_no)_satisfaction.txt')


    

Searching for songs by The Beatles...

Song 1: "Yesterday"

Reached user-specified song limit (1).
Done. Found 1 songs.
Searching for "Hey Jude" by The Beatles...
Done.
Searching for songs by Led Zeppelin...

Song 1: "Stairway to Heaven"

Reached user-specified song limit (1).
Done. Found 1 songs.
Searching for songs by Pink Floyd...

Song 1: "Wish You Were Here"

Reached user-specified song limit (1).
Done. Found 1 songs.
Searching for "Another Brick in the Wall" by Pink Floyd...
Done.
Searching for songs by The Rolling Stones...

Song 1: "Sympathy for the Devil"

Reached user-specified song limit (1).
Done. Found 1 songs.
Searching for "(I Cant Get No) Satisfaction" by The Rolling Stones...
Done.


4. Write a function that takes the name of a file that contains song lyrics, loads the file, performs sentiment analysis, and returns the polarity score.  Use this function to print the polarity scores (with the name of the song) of the three files you created in question 3.  Does the reported polarity match your understanding of the song's lyrics? Why or why not do you think that might be?  Answer the questions in either a comment in the code cell or a markdown cell under the code cell.

In [72]:
import json
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

def lyric_polarity (filename):
    nlp=spacy.load('en_core_web_sm')
    nlp.add_pipe('spacytextblob')
    with open(filename, 'r') as f:
        data = json.load(f)
    doc = nlp(data)
    polarity = doc._.blob.polarity 

    print(f"Polarity of {filename}: {polarity}")

lyric_polarity("hey_jude.txt_lyrics.json")
lyric_polarity("(i_cant_get_no)_satisfaction.txt_lyrics.json")
lyric_polarity("another_brick_in_the_wall.txt_lyrics.json")
lyric_polarity("stairway_to_heaven.txt_lyrics.json")

Polarity of hey_jude.txt_lyrics.json: 0.0502777777777778
Polarity of (i_cant_get_no)_satisfaction.txt_lyrics.json: 0.0359375
Polarity of another_brick_in_the_wall.txt_lyrics.json: 0.0
Polarity of stairway_to_heaven.txt_lyrics.json: 0.09801925505050504


"The polarity score of hey_jude.txt_lyrics.json is 0.0502777777777778, which indicates a slightly positive sentiment. The song "Hey Jude" is about giving someone hope and encouragement, so it makes sense that the polarity score is slightly positive.

The polarity score of (i_cant_get_no)_satisfaction.txt_lyrics.json is 0.0359375, which indicates a slightly negative sentiment. The song "(I Can't Get No) Satisfaction" is about dissatisfaction with the status quo, so it makes sense that the polarity score is slightly negative.

The polarity score of another_brick_in_the_wall.txt_lyrics.json is 0.0, which indicates a neutral sentiment. The song "Another Brick in the Wall" is about alienation and conformity, so it makes sense that the polarity score is neutral.

The polarity score of stairway_to_heaven.txt_lyrics.json is 0.09801925505050504, which indicates a slightly positive sentiment. The song "Stairway to Heaven" is about spirituality and transcendence, so it makes sense that the polarity score is slightly positive."